### PR TITLE
Add support of RESP3 verbatim string

### LINKED ISSUE #2036: Add RESP3 verbatim string type

### PR DESCRIPTION
**Pull request by wsehjk**
This pr is to close #2036.

### CODE REVIEW

**Comment by wsehjk** [comment]
This implementation refers other RESPV3 type support, I'm really not sure whether it works. Plz review my code. Thanks

File: kvrocks.conf [hunk 1]
```diff
@@ -912,3 +912,5 @@ rocksdb.rate_limiter_auto_tuned yes
 
 ################################ NAMESPACE #####################################
 # namespace.test change.me
+backup-dir /tmp/kvrocks/backup
+pidfile /tmp/kvrocks/kvrocks.pid
```
Comment (PragmaTwice) [inline kvrocks.conf]: Please revert this change.

File: src/commands/cmd_server.cc [hunk 2]
```diff
@@ -648,16 +648,18 @@ class CommandDebug : public Commander {
             redis::BulkString("key:123"),
             redis::Integer(90),
         });
+      } else if (protocol_type_ == "verbatim") {  // verbatim string
+        *output = conn->VerbatimString("txt", "verbatim string");
       } else {
         *output = redis::Error(
             "Wrong protocol type name. Please use one of the following: "
-            "string|integer|double|array|set|bignum|true|false|null|attrib");
+            "string|integer|double|array|set|bignum|true|false|null|attrib|verbatim_string");
```
Comment (PragmaTwice) [inline src/commands/cmd_server.cc]:
```suggestion
            "string|integer|double|array|set|bignum|true|false|null|attrib|verbatim");
```

File: src/server/redis_connection.h [hunk 3]
```diff
@@ -71,6 +71,12 @@ class Connection : public EvbufCallbackBase<Connection> {
   std::string Double(double d) const {
     return protocol_version_ == RESP::v3 ? "," + util::Float2String(d) + CRLF : BulkString(util::Float2String(d));
   }
+  // type should be "txt" or "mkd"
+  // if RESP is V2, treat verbatim string as blob string
+  std::string VerbatimString(const std::string &type, const std::string &data) const {
+    return protocol_version_ == RESP::v3 ? "=" + std::to_string(3 + data.size()) + CRLF + type + ":" + data + CRLF
+                                         : BulkString(data);
+  }
```
Comment (PragmaTwice) [inline src/server/redis_connection.h]:
```suggestion
  std::string VerbatimString(const std::string &type, const std::string &data) const {
    return protocol_version_ == RESP::v3 ? "=" + std::to_string(type.size() + data.size()) + CRLF + type + ":" + data + CRLF
                                         : BulkString(data);
  }
```
Even the length of `type` *should* be 3, but there is no check so if we pass a wrong type it will be a big issue.

**Comment by git-hulk** [comment]
@wsehjk You can search [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code) inside Redis codebase to see where's using the verbatim string and also replace it in Kvrocks.

**Comment by git-hulk** [comment]
@wsehjk You can search [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code) inside Redis codebase to see where's using the verbatim string and also replace it in Kvrocks.

**Comment by PragmaTwice** [comment]
Repeated comment deleted.

File: src/server/redis_connection.h (hunk 3 above)
Comment (wsehjk) [reply src/server/redis_connection.h]: I think this function should check `type` is 'txt' or 'mkd', but i don't know how to handle the wrong type. Throw an error??

Comment (PragmaTwice) [reply src/server/redis_connection.h]: Currently I think you don't need to check. Or you can use the `CHECK` macro from glog.

Comment (wsehjk) [reply src/server/redis_connection.h]: Ok, thanks

**Comment by wsehjk** [comment]
https://github.com/wsehjk/kvrocks/actions/runs/7932113585.  CI has passed

File: src/server/redis_connection.h [hunk 4]
```diff
@@ -71,6 +71,17 @@ class Connection : public EvbufCallbackBase<Connection> {
   std::string Double(double d) const {
     return protocol_version_ == RESP::v3 ? "," + util::Float2String(d) + CRLF : BulkString(util::Float2String(d));
   }
+  // ext is the extension of file to send, 'txt' for text file, 'md ' for markdown file
+  // at most 3 chars, padded with space
+  // if RESP is V2, treat verbatim string as blob string
+  // https://github.com/redis/redis/blob/7.2/src/networking.c#L1099
+  std::string VerbatimString(std::string ext, const std::string &data) const {
+    CHECK(ext.size() <= 3);
+    int padded_len = 3 - static_cast<int>(ext.size());
```
Comment (PragmaTwice) [inline src/server/redis_connection.h]:
```suggestion
    size_t padded_len = 3 - ext.size();
```

File: src/server/redis_connection.h [hunk 5]
```diff
@@ -71,6 +71,17 @@ class Connection : public EvbufCallbackBase<Connection> {
...
+    ext = ext + std::string(padded_len, ' ');
+    return protocol_version_ == RESP::v3 ? "=" + std::to_string(3 + data.size() + 1) + CRLF + ext + ":" + data + CRLF
```
Comment (PragmaTwice) [inline src/server/redis_connection.h]:
```suggestion
    return protocol_version_ == RESP::v3 ? "=" + std::to_string(3 + 1 + data.size()) + CRLF + ext + ":" + data + CRLF
```

File: src/server/redis_connection.h (hunk 4 above)
Comment (wsehjk) [reply src/server/redis_connection.h]: Ok, `size_t` is more natural as std::string() argument type

File: src/server/redis_connection.h (hunk 5 above)
Comment (wsehjk) [reply src/server/redis_connection.h]: `3+1+data.size()` describes the format `ext:data` better. thanks

**Comment by wsehjk** [comment]
Code updated.

**Comment by PragmaTwice** [comment]
> @wsehjk You can search [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code) inside Redis codebase to see where's using the verbatim string and also replace it in Kvrocks.

@wsehjk As mentioned by @git-hulk , besides adding the reply function, we need to replace it in existing command impls.

**Comment by wsehjk** [comment]
> > @wsehjk You can search [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code) inside Redis codebase to see where's using the verbatim string and also replace it in Kvrocks.
> 
> @wsehjk As mentioned by @git-hulk , besides adding the reply function, we need to replace it in existing command impls.

Well, I figure out what you mean, wan't ware of this additional work before. I will work on this. But can we do this in a seperate issue and merge this pr #2102?

**Comment by git-hulk** [comment]
@wsehjk After taking a look at [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code), only one place that needs to be changed in Kvrocks: use verbatim string while replying to the cluster command: https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/cluster.c#L853

And for the Lua part, I will do it in the next PR.

**Comment by wsehjk** [comment]
> @wsehjk After taking a look at [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code), only one place that needs to be changed in Kvrocks: use verbatim string while replying to the cluster command: https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/cluster.c#L853
> 
> And for the Lua part, I will do it in the next PR.

what about the `info` cmd and `client` cmd https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/server.c#L6082
https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/networking.c#L3086

**Comment by git-hulk** [comment]
> > @wsehjk After taking a look at [addReplyVerbatim](https://github.com/search?q=repo%3Aredis%2Fredis%20addReplyVerbatim&type=code), only one place that needs to be changed in Kvrocks: use verbatim string while replying to the cluster command: https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/cluster.c#L853
> > And for the Lua part, I will do it in the next PR.
> 
> what about the `info` cmd and `client` cmd https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/server.c#L6082 https://github.com/redis/redis/blob/c854873746e9515f322d805f863e839ebfcefd62/src/networking.c#L3086

Yes, can fix them as well.

**Comment by wsehjk** [comment]
As discussed before, I replace `redis::BulkString` with `connection->VerbatimString()`reply in `info`, `client` and `cluster`command where in redis codebae `addReplyVerbatim` reply is used

**Review by git-hulk — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

**Comment by wsehjk** [comment]
Thanks

**Comment by git-hulk** [comment]
@wsehjk Thanks for your contributions!
